File: Codeforces/Div 3/Codeforces Round 998/2060E.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
#define endl '\n'

class UnionFind {
public:
    vector<int> pa, sz;
    int cnt;

    UnionFind(int n) {
        pa.resize(n);
        iota(pa.begin(), pa.end(), 0);
        sz.assign(n, 1);
        cnt = n;
    }

    int find(int x) {
        while (pa[x] != x) {
            pa[x] = pa[pa[x]];
            x = pa[x];
        }
        return x;
    }

    bool unionSet(int x, int y) {
        x = find(x), y = find(y);
        if (x == y) return false;
        if (sz[x] < sz[y]) swap(x, y);
        pa[y] = x;
        sz[x] += sz[y];
        cnt--;
        return true;
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int t, n, m1, m2;
    cin >> t;
    while (t--) {
        cin >> n >> m1 >> m2;
        vector<pair<int,int>> edges1(m1), edges2(m2);
        for (auto& [u, v] : edges1) {
            cin >> u >> v;
            u--, v--;
        }
        for (auto& [u, v] : edges2) {
            cin >> u >> v;
            u--, v--;
        }

        LL ans = 0;
        UnionFind uf1(n), uf2(n);
        for (auto [u, v] : edges2) {
            uf2.unionSet(u, v);
        }
        for (auto [u, v] : edges1) {
            if (uf2.find(u) != uf2.find(v)) {
                ans++;
            } else {
                uf1.unionSet(u, v);
            }
        }

        vector<pair<int,int>> pairs(n);
        for (int i = 0; i < n; i++) {
            pairs[i] = {uf2.find(i), uf1.find(i)};
        }
        sort(pairs.begin(), pairs.end());

        int i = 0;
        while (i < n) {
            int cnt = 1;
            int prev1 = pairs[i].second;
            int curr2 = pairs[i].first;
            i++;
            while (i < n && pairs[i].first == curr2) {
                if (pairs[i].second != prev1) {
                    cnt++;
                    prev1 = pairs[i].second;
                }
                i++;
            }
            if (cnt > 1) {
                ans += (cnt - 1);
            }
        }
        cout << ans << endl;
    }
    return 0;
}

File: Codeforces/Div 3/Codeforces Round 998/2060E.py
from collections import *

class UnionFind:
    __slots__ = ['n', 'pa', 'sz', 'cnt']

    def __init__(self, n: int):
        self.n = n
        self.pa = list(range(n)) # 父節點
        self.sz = [1] * n # 連通分量大小
        self.cnt = n # 連通分量數量

    def find(self, x: int) -> int:
        while self.pa[x] != x:
            self.pa[x] = self.pa[self.pa[x]]
            x = self.pa[x]
        return x
    
    def union(self, x: int, y: int) -> bool:
        px, py = self.find(x), self.find(y)
        if px == py:
            return False
        if self.sz[px] < self.sz[py]:
            px, py = py, px
        self.pa[py] = px
        self.sz[px] += self.sz[py]
        self.cnt -= 1
        return True

t = int(input())
for _ in range(t):
    n, m1, m2 = map(int, input().split())
    edges1 = []
    for _ in range(m1):
        u, v = map(lambda x: int(x) - 1, input().split())
        edges1.append((u, v))
    edges2 = []
    for _ in range(m2):
        u, v = map(lambda x: int(x) - 1, input().split())
        edges2.append((u, v))

    uf2 = UnionFind(n)
    for u, v in edges2:
        uf2.union(u, v)

    ans = 0
    uf1 = UnionFind(n)
    for (u, v) in edges1:
        if uf2.find(u) != uf2.find(v):  # 刪除這條邊
            ans += 1
        else:
            uf1.union(u, v)

    # 將 G 的連通分量分組
    comps = defaultdict(list)
    for u in range(n):
        comps[uf2.find(u)].append(u)

    for _, nodes in comps.items():
        # G 的這個連通分量，在 F 中被分成了 len(nodes) 個連通分量，需要將其連接起來
        roots = set()
        for u in nodes:
            roots.add(uf1.find(u))
        cost = len(roots)
        if cost > 1:
            ans += (cost - 1)

    print(ans)